avalanche/training/templates/observation_type/batch_observation.py

```diff
@@ -22,36 +43,29 @@ def model_adaptation(self, model=None):
             # If the strategy has access to task boundaries, adapt the model
             # for the whole origin experience to add the
             if self.experience.access_task_boundaries:
-                avalanche_model_adaptation(model,
+                avalanche_model_adaptation(model, 
                                            self.experience.origin_experience)
             else:
                 avalanche_model_adaptation(model, self.experience)
-
-        # For evaluation, the experience is not necessarily an online
-        # experience:
         else:
             avalanche_model_adaptation(model, self.experience)
 
         return model.to(self.device)
 
-    def make_optimizer(self):
+    def make_optimizer(self, reset_opt=True, reset_state=False):
         """Optimizer initialization.
 
         Called before each training experiene to configure the optimizer.
         """
-        if isinstance(self.experience, OnlineCLExperience):
-            if self.experience.access_task_boundaries:
-                assert self.experience.is_first_subexp
-                reset_optimizer(self.optimizer, self.model)
+        reset_opt = detect_param_change(
+            self.optimizer, list(self.model.named_parameters()),
+        ) or reset_opt
```

Comment:
this is different from what we discussed. Only the dynamic modules should be reset unless `reset_opt` is set. The state of the others should be kept.

Reply:
you can check the shape of parameters before/after adaptation. If they changed, you should reset the state.

Reply:
> In other words, I know how to reset the whole state, but not one part of the state only.

I don't know how to fix this honestly. This is why we just used to reset everything.

> One way I see would be to have one parameter group per parameter but this would be uncompatible with user defined parameter groups

I would not mess with parameter groups since they may be defined by the user.

Reply:
> you can check the shape of parameters before/after adaptation. If they changed, you should reset the state for those parameters.

This is already what I am doing, the id of the parameter changes whenever the id changes. I could check the shape also but this would be redundant I think. Though I should make sure that this thing (id change = shape change) is always true, but I think it is. I might have found a way to update the state btw, Ill push smt today.